km/km_filesys.c

```diff
@@ -2624,10 +2677,46 @@ static inline int km_fs_recover_socket(km_nt_socket_t* nt_sock, struct sockaddr*
    km_file_t* file = &km_fs()->guest_files[nt_sock->fd];
    km_assert(km_is_file_used(file) == 0);
 
-   int host_fd = socket(nt_sock->domain, nt_sock->type, nt_sock->protocol);
-   if (host_fd < 0) {
-      km_warn("socket recover");
-      return -1;
+   int host_fd;
+   // check if this is lightweight snap listening socket, retore it accordingly
+   if (snap_listen_sockinfo != NULL && nt_sock->domain == snap_listen_sockinfo->domain &&
+       memcmp(nt_sock + 1, snap_listen_sockinfo->addr, snap_listen_sockinfo->addrlen)) {
```

Comment:
Should this be memcmp(....) == 0?
Don't you want to check for a matching bind address?
I guess since the domains are the same we don't need to compare  nt_sock->addrlen with snap_listen_sockinfo->addrlen for equality.

Reply:
I guess you can listen again on the same port + addr and the kernel recognizes 2 listeners on different fd's for the same address.

Reply:
SO_REUSEADDR helped too I think.